pkg/streamingpromql/optimize/ast/reduce_matchers.go

```diff
@@ -1,62 +1,100 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 
-package lookupplan
+package ast
 
 import (
 	"context"
 
-	"github.com/grafana/dskit/tracing"
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
 	"github.com/prometheus/prometheus/model/labels"
-	"github.com/prometheus/prometheus/tsdb/index"
-	"go.opentelemetry.io/otel/attribute"
-	"go.opentelemetry.io/otel/trace"
+	"github.com/prometheus/prometheus/promql/parser"
 
 	"github.com/grafana/mimir/pkg/util"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
 )
 
-// MatcherReducerPlanner deduplicates matchers from the input plan,
-// and removes matchers that select for all non-empty values if a more selective matcher for the same label name already exists.
-// It does not modify plans if the input plan has any scan matchers, and does not guarantee consistent matcher ordering in the output plan.
-type MatcherReducerPlanner struct{}
-
-// concreteLookupPlan implements LookupPlan by storing pre-computed index and scan matchers.
-type concreteLookupPlan struct {
-	indexMatchers []*labels.Matcher
-	scanMatchers  []*labels.Matcher
+func NewReduceMatchers(reg prometheus.Registerer, logger log.Logger) *ReduceMatchers {
+	return &ReduceMatchers{
+		attempts: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_mimir_query_engine_reduce_matchers_attempted_total",
+			Help: "Total number of queries that the optimization pass has attempted to reduce matchers for.",
+		}),
+		success: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_mimir_query_engine_reduce_matchers_success_total",
```

Comment:
[nit] "success" might be ambiguous or confusing here, as this counter not being incremented doesn't imply that the optimisation pass failed with an error. Maybe "expressions modified" or something like that would be clearer?

Reply:
Fixed in d5d4ae7f0b72436f5f59cc0a6f786f32222d142b